File: backend/test_device_flow.js
const { execSync } = require('child_process');

async function run() {
  try {
    const BASE_URL = "http://localhost:3000/api";
    const email = `test_${Date.now()}@test.com`;
    
    console.log("Registering user...");
    const resAuth = await fetch(`${BASE_URL}/auth/register`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ name: "Test User", email, password: "password123" })
    });
    const authData = await resAuth.json();
    let token = authData.token;
    
    if (!token) {
        const resLogin = await fetch(`${BASE_URL}/auth/login`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ email, password: "password123" })
        });
        const loginData = await resLogin.json();
        token = loginData.token;
    }
    
    const headers = { 'Authorization': `Bearer ${token}`, 'Content-Type': 'application/json' };

    console.log("Creating workspace...");
    const resWs = await fetch(`${BASE_URL}/workspaces`, {
        method: 'POST',
        headers,
        body: JSON.stringify({ name: "My Workspace" })
    });
    const wsData = await resWs.json();
    const workspaceId = wsData.workspace.id;

    console.log("Creating device...");
    const resDev = await fetch(`${BASE_URL}/workspaces/${workspaceId}/devices`, {
        method: 'POST',
        headers,
        body: JSON.stringify({ name: "Test Sensor", device_type: "ESP32" })
    });
    const devData = await resDev.json();
    const deviceId = devData.device.device_id;
    const secretKey = devData.device.secret_key;
    const deviceUuid = devData.device.id;

    console.log(`Created Device: ${deviceId} with Secret: ${secretKey}`);

    console.log("Running simulator...");
    const output = execSync(`python3 ../scripts/test_mqtt_device.py --device-id ${deviceId} --secret-key ${secretKey} --count 3 --interval 1`).toString();
    console.log(output);

    console.log("Checking device status...");
    const resCheck = await fetch(`${BASE_URL}/workspaces/${workspaceId}/devices/${deviceUuid}`, { headers });
    const checkData = await resCheck.json();
    const status = checkData.device.status;
    const lastSeen = checkData.device.last_seen;
    
    console.log(`Backend Device Status: ${status}`);
    console.log(`Backend Last Seen: ${lastSeen}`);

    if (status.toLowerCase() !== 'online') {
        console.error("Status did not update to online!");
        process.exit(1);
    }
    console.log("E2E Test Passed!");
  } catch (err) {
    console.error(err);
    process.exit(1);
  }
}
run();

File: test_device_flow.py
import requests
import json
import subprocess
import time

BASE_URL = "http://localhost:3000/api"

# 1. Register User
email = f"test_{int(time.time())}@test.com"
res = requests.post(f"{BASE_URL}/auth/register", json={"name": "Test User", "email": email, "password": "password123"})
if res.status_code != 201:
    print("User creation failed")
    exit(1)
token = res.json()['token']
headers = {"Authorization": f"Bearer {token}"}

# 2. Create Workspace
res = requests.post(f"{BASE_URL}/workspaces", json={"name": "My Workspace"}, headers=headers)
workspace_id = res.json()['workspace']['id']

# 3. Create Device
res = requests.post(f"{BASE_URL}/workspaces/{workspace_id}/devices", json={"name": "Test Sensor", "device_type": "ESP32"}, headers=headers)
device_data = res.json()['device']
device_id = device_data['device_id']
secret_key = device_data['secret_key']
device_uuid = device_data['id']

print(f"Created Device: {device_id} with Secret: {secret_key}")

# 4. Run Simulator
print("Running simulator...")
cmd = ["python3", "scripts/test_mqtt_device.py", "--device-id", device_id, "--secret-key", secret_key, "--count", "3"]
result = subprocess.run(cmd, capture_output=True, text=True)
print(result.stdout)
if result.returncode != 0:
    print(result.stderr)
    exit(1)

# 5. Check Device Status
res = requests.get(f"{BASE_URL}/workspaces/{workspace_id}/devices/{device_uuid}", headers=headers)
status = res.json()['device']['status']
last_seen = res.json()['device']['last_seen']
print(f"Backend Device Status: {status}")
print(f"Backend Last Seen: {last_seen}")

if status != 'ONLINE' and status != 'online':
    print("Status did not update to online!")
    exit(1)

print("E2E Test Passed!")
